### PR TITLE
Improve `TryDecodeEntireState` output

### PR DESCRIPTION
**Pull request by liamaharon**
Found some areas for improvement while working on https://github.com/polkadot-fellows/runtimes/pull/122.

## Improvements

- If multiple keys in a storage item (e.g. a map) are undecodable, return all the undecodable keys rather than only the first one found 
- Include the key of the undecodable storage in the INFO log
- Write output as hex string where appropriate
- Write INFO log on successful decoding

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4738438

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4739540

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4739539

File: substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs [hunk 1]
```diff
@@ -104,29 +117,38 @@ fn decode_storage_info<V: Decode>(
...
-				}]
+				}
 			})?;
 
-			Ok::<usize, Vec<_>>(len)
+			Ok::<usize, _>(len)
 		},
 	};
 
-	decoded += decode_key(&next_key)?;
```
Comment (liamaharon) [inline substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs:118]: todo: not miss this key

File: substrate/frame/support/Cargo.toml [hunk 2]
```diff
@@ -34,6 +34,7 @@ sp-weights = { path = "../../primitives/weights", default-features = false }
 sp-debug-derive = { path = "../../primitives/debug-derive", default-features = false }
 sp-metadata-ir = { path = "../../primitives/metadata-ir", default-features = false }
 tt-call = "1.0.8"
+hex = { default-features = false, version = "0.4.3", features = ["alloc"] }
```
Comment (bkchr) [inline substrate/frame/support/Cargo.toml:37]: The crate array-bytes already supports this.

File: substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs (hunk 1 above)
Comment (bkchr) [reply substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs:118]: What you mean by this to do?

Comment (liamaharon) [reply substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs:118]: I realised it currently skips the first key. Will fix today.

Comment (liamaharon) [reply substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs:118]: fixed

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4743805

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4743806

File: substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs [hunk 3]
```diff
@@ -96,37 +109,46 @@ impl core::fmt::Display for TryDecodeEntireStorageError {
...
+	let mut errors = vec![];
+	let mut next_key = Some(info.prefix.clone());
```
Comment (bkchr) [inline substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs:131]: We should just use `PrefixKeyIterator` here. Not sure why we have written this here "manually" :P

Comment (liamaharon) [reply substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs:131]: Disadvantage of `KeyPrefixIterator` is that it "skips" keys that error. We want to handle the Result type ourselves.

**Review by ggwpez — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4767583

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4767584

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4767585

File: substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs (hunk 1 above)
Comment (bkchr) [reply substrate/frame/support/src/traits/try_runtime/decode_entire_state.rs:118]: BTW, we did not miss the key. The first key being the `prefix` and we don't store there anything, but given how the code is written there should also be no problem, just one more db access.

**Review by bkchr — APPROVED** [review]

**Pull request merged**